test/testdata/resolver/final_module.rb

```diff
@@ -0,0 +1,58 @@
+# typed: true
+# disable-fast-path: true
+
+class FinalClass
+  extend T::Helpers
+  final!
+end
+
+module FinalModule
+  extend T::Helpers
+  final!
+end
+
+class BadInherit < FinalClass # error: `FinalClass` was declared as final and cannot be inherited by `BadInherit`
+end
+
+module BadInclude # error: `FinalModule` was declared as final and cannot be included by `BadInclude`
+  include FinalModule
+end
+
+module BadExtend # error: `FinalModule` was declared as final and cannot be extended by `BadExtend`
```

Comment:
You can use some information on the `ClassDef` to give a better loc if you want:

https://github.com/sorbet/sorbet/blob/0a31f41f57cd185f87c85f5cc578fdb02424159a/ast/Trees.h#L104